### PR TITLE
Add copy_triangulation to parallel::fullydistributed::Triangulation

### PR DESCRIPTION
**Pull request by peterrum**
While integrating `p:f:t` in a user code, I noticed it is sometimes somewhat annoying to pass the information how to partition a serial triangulation and how to construct the `ConstructionData`. Instead, I propose in this PR *register* the information in `p:f:t`, which is used during a new `copy_triangulation`-method to construct the `ConstrucionData`.

An example (part of the new test `fullydistributed_grids/copy_serial_tria_07`, which is a variant of  `fullydistributed_grids/copy_serial_tria_02`):

```cpp
  // create instance of pft
  parallel::fullydistributed::Triangulation<dim> tria_pft(comm);

  // register dynamic construction data
  tria_pft.register_dynamic_construction_data(
    std::make_shared<
      const parallel::fullydistributed::DynamicConstructionData<dim>>(
      [](dealii::Triangulation<dim> &tria, const unsigned int n_partitions) {
        GridTools::partition_triangulation_zorder(n_partitions, tria);
      },
      true));

  // actually create triangulation
  tria_pft.copy_triangulation(basetria);
```
@bangerth @tjhei Your opinion would be valuable! I think in a similar way we might want to handle re-partitioning with arbitrary *oracles* in the future...

ping @eliasstudiert

### CODE REVIEW

**Comment by peterrum** [comment]
Note: the Doxygen documentation is still missing...

**Comment by peterrum** [comment]
I have added the doxygen documentation. Ready to be reviewed!

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 1]
```diff
@@ -322,6 +370,31 @@ namespace parallel
                            const std::vector<dealii::CellData<dim>> &cells,
                            const SubCellData &subcelldata) override;
 
+      /**
+       * Implementation of the same function as in the base class.
+       *
+       * @param other_tria The triangulation to be copied. It can be a serial
+       *        Triangulation or a parallel::distributed::Triangulation. Both
+       *        can have already been refined.
```
Comment (kronbichler) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
       *        can have been refined already.
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 2]
```diff
@@ -322,6 +370,31 @@ namespace parallel
...
+       *
+       * @note Please note that this function requires that a
+       *       DynamicConstructionData has been registered, which contains
+       *       the configuration how to create the ConstructionData that
```
Comment (kronbichler) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
       *       the configuration how to create the ConstructionData to be
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 3]
```diff
@@ -322,6 +370,31 @@ namespace parallel
...
+       *       is passed automatically to create_triangulation.
```
Comment (kronbichler) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
       *       passed automatically to create_triangulation.
```

File: source/distributed/fully_distributed_tria.cc [hunk 4]
```diff
@@ -285,6 +289,98 @@ namespace parallel
...
 
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::copy_triangulation(
+      const dealii::Triangulation<dim, spacedim> &other_tria)
+    {
+      Assert(this->dynamic_construction_data,
+             ExcMessage("No DynamicConstructionData has been provided!"));
+
+      // temporal array to store user subdomain_ids and level_subdomain_ids
```
Comment (kronbichler) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
      // temporary array to store user subdomain_ids and level_subdomain_ids
```
?

File: source/distributed/fully_distributed_tria.cc [hunk 5]
```diff
@@ -285,6 +289,98 @@ namespace parallel
...
+      // (since these might be changed during partitioning)
+      std::vector<std::pair<types::subdomain_id, types::subdomain_id>>
+        stored_cell_subdomain_ids;
+
+      if (dynamic_cast<const dealii::parallel::TriangulationBase<dim, spacedim>
+                         *>(&other_tria) == nullptr)
+        {
+          // check if other triangulation is not parallel triangulation;
+          // this means we have a serial triangulation, which has to be
+          // partitioned!
```
Comment (kronbichler) [inline source/distributed/fully_distributed_tria.cc]: Maybe move this above the if statement? I would write starting from the middle of the first line `... is not a parallel one, which needs to be partitioned`

File: source/distributed/fully_distributed_tria.cc [hunk 6]
```diff
@@ -285,6 +289,98 @@ namespace parallel
...
+
+          auto &serial_tria =
+            *const_cast<dealii::Triangulation<dim, spacedim> *>(&other_tria);
```
Comment (kronbichler) [inline source/distributed/fully_distributed_tria.cc]: It seems that you need the `const_cast` to be compatible with the various `GridTools::partition` functions. From an interface perspective, would it make sense to either copy the triangulation here? I know that you restore the old state in the end, but the question is whether the memory consumption is significant. If yes, please add a comment along those lines ("we rather overwrite the given triangulation's subdomain ids here than create a more memory-heavy copy").

File: source/distributed/fully_distributed_tria.cc [hunk 7]
```diff
@@ -285,6 +289,71 @@ namespace parallel
...
+      // if the source triangulation is serial
+      dealii::Triangulation<dim, spacedim> serial_tria;
+
+      // check if other triangulation is not a parallel one, which needs to be
+      // partitioned
+      if (dynamic_cast<const dealii::parallel::TriangulationBase<dim, spacedim>
+                         *>(&other_tria) == nullptr)
+        {
+          // actually copy the serial triangulation
+          serial_tria.copy_triangulation(
```
Comment (peterrum) [inline source/distributed/fully_distributed_tria.cc]: @kronbichler  Is this more after your liking?

**Comment by peterrum** [comment]
@kronbichler Thanks for the review!

File: source/distributed/fully_distributed_tria.cc (hunk 7 above)
Comment (kronbichler) [reply source/distributed/fully_distributed_tria.cc]: Yes, I think this is much simpler. My only question was whether there was a cost/memory consumption reason - but I guess we can tolerate that here.

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 8]
```diff
@@ -208,6 +210,53 @@ namespace parallel
...
+        std::function<void(dealii::Triangulation<dim, spacedim> &,
+                           const unsigned int)> partitioner =
+          [](dealii::Triangulation<dim> &tria,
+             const unsigned int          n_partitions) {
+            GridTools::partition_triangulation_zorder(n_partitions, tria);
+          },
+        bool construct_multilevel_hierarchy = false)
+        : partitioner(partitioner)
+        , construct_multilevel_hierarchy(construct_multilevel_hierarchy)
+      {}
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: Can you move the implementation to a `.cc` file?

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 9]
```diff
@@ -208,6 +210,53 @@ namespace parallel
...
+     */
+    template <int dim, int spacedim = dim>
+    struct DynamicConstructionData
+    {
+      /**
+       * Constructor.
+       *
+       * @note As a default, GridTools::partition_triangulation_zorder() is used
+       *       as partitioner and data structures on multigrid levels are not
+       * set up.
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: indentation issue here.

I have to admit that from neither the class nor the constructor documentation do I understand what this class does. It has a reference to `pft::copy_triangulation()`, but I think the documentation would be useful here.

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 10]
```diff
@@ -208,6 +210,53 @@ namespace parallel
...
+       */
+      DynamicConstructionData(
+        std::function<void(dealii::Triangulation<dim, spacedim> &,
+                           const unsigned int)> partitioner =
+          [](dealii::Triangulation<dim> &tria,
+             const unsigned int          n_partitions) {
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: It would be useful if you could document what the arguments of the function argument are.

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 11]
```diff
@@ -208,6 +210,53 @@ namespace parallel
...
+            GridTools::partition_triangulation_zorder(n_partitions, tria);
+          },
+        bool construct_multilevel_hierarchy = false)
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
        const bool construct_multilevel_hierarchy = false)
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 12]
```diff
@@ -208,6 +210,53 @@ namespace parallel
...
+       * of the function) in n partitions (second parameter of the function).
+       */
+      std::function<void(dealii::Triangulation<dim, spacedim> &,
+                         const unsigned int)>
+        partitioner;
+
+      /**
+       * Signal if the multigrid levels should be constructed.
+       */
+      bool construct_multilevel_hierarchy;
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
      const bool construct_multilevel_hierarchy;
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 13]
```diff
@@ -322,6 +371,31 @@ namespace parallel
...
                            const SubCellData &subcelldata) override;
 
+      /**
+       * Implementation of the same function as in the base class.
+       *
+       * @param other_tria The triangulation to be copied. It can be a serial
+       *        Triangulation or a parallel::distributed::Triangulation. Both
+       *        can have been refined already.
+       *
+       * @note Please note that this function requires that a
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: I think we overuse "Please note", especially if it's preceded by `@note` :-)
```suggestion
       * @note This function requires that a
```

File: source/distributed/fully_distributed_tria.cc [hunk 14]
```diff
@@ -285,6 +289,71 @@ namespace parallel
...
+      // pointer to the triangulation for which the construction data
+      // should be created (normally it is the input triangulation but
+      // in the case of a serial triangulation we create a copy which should
+      // be used)
+      const dealii::Triangulation<dim, spacedim> *other_tria_ptr = &other_tria;
+
+      // temporary serial triangulation (since we might modifying subdomain_ids
+      // and level_subdomain_ids during partitioning); it is filled only
+      // if the source triangulation is serial
+      dealii::Triangulation<dim, spacedim> serial_tria;
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]: I'd make that a `std::unique_ptr` so that we don't allocate any memory if you don't need the object.

Comment (bangerth) [reply source/distributed/fully_distributed_tria.cc]: can you say why you actually need the variable at all? I suspect because you partition the serial triangulation and you can't do that on the input argument?

Comment (bangerth) [reply source/distributed/fully_distributed_tria.cc]: ...since it's `const`?

File: source/distributed/fully_distributed_tria_util.inst.in [hunk 15]
```diff
@@ -15,34 +15,36 @@
 
 
 
-for (deal_II_dimension : DIMENSIONS)
+for (deal_II_dimension : DIMENSIONS; deal_II_spacedim : DIMENSIONS)
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria_util.inst.in]: I think that in most places we actually use this spelling:
```suggestion
for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension : DIMENSIONS)
```

**Comment by peterrum** [comment]
@bangerth Thanks for the review. I have updated the code accordingly.

However, you misunderstood the usage of the new function (which was probably related to the documentation):

> I think that your approach is overly complicated because you first need to create a struct that doesn't actually do anything other than store two variables, then you need to attach that to the pft, and then it's used in copy_triangulation(). I think that you do that because you feel constrained by the interface of the virtual copy_triangulation() function. But if the interface doesn't fit, then don't use it!

I have created the two functions `register_dynamic_construction_data` and `copy_triangulation` intentionally (and not for the reasons you a describing). I want to decide during initialization which partitioner I want to use (`p4est` or `METIS`) and  call  `register_dynamic_construction_data` only once for `p:f:t` and after then treat `p:f:t` as a regular `Triangulation`. The registered information (including the partioner) are used during `copy_triangulation` and will hopefully used in the future for AMR (in that case you do not want to pass a partitioner every time you refine/coarsen the mesh). 

In other words if we consider `parallel::distributed::Triangulation`: In the case of `p:d:t`,  `register_dynamic_construction_data` is *called* in the constructor with a **fixed** partitioner (p4est). After that this partitioner is uses during AMR. This close relationship of `p:d:t` and `p4est` is something @tjhei (see https://github.com/dealii/dealii/pull/8418#issuecomment-515769415) and I do not like.

This is a step towards actually separating concerns of a triangulation and a partitioning class!

ping @tjhei

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 16]
```diff
@@ -208,6 +210,63 @@ namespace parallel
...
+    template <int dim, int spacedim = dim>
+    struct DynamicConstructionData
+    {
+      /**
+       * Constructor.
+       *
+       * @param partitioner A partitioning function, which takes as input argument
+       *                    a reference to the triangullation to be partitioned
+       * and the number of partitions to be created.
+       * @param construct_multilevel_hierarchy Signal if the multigrid levels
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: Indent line 237

**Comment by peterrum** [comment]
@bangerth  Now I am passing a partitioner lambda directly to the function.

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 17]
```diff
@@ -322,6 +325,39 @@ namespace parallel
...
+      void
+      copy_triangulation(
+        const dealii::Triangulation<dim, spacedim> &other_tria) override;
+
+      /**
+       * Register a partitioner, which is used within the method
+       * copy_triangulation.
+       *
+       * @param partitioner A partitioning function, which takes as input argument
+       *                    a reference to the triangullation to be partitioned
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
       *                    a reference to the triangulation to be partitioned
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 18]
```diff
@@ -322,6 +325,39 @@ namespace parallel
...
+       *                    and the number of partitions to be created.
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: Can you say what the function is supposed to do? I suspect that it is supposed to set subdomain ids? And that the triangulation it gets is a copy of a triangulation that can be changed freely. Maybe like this:
```suggestion
       *                    and the number of partitions to be created. The function needs to set subdomain
       *                    ids for each active cell of the given triangulation, with values between zero (inclusive)
       *                    and the second argument to the function (exclusive).
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 19]
```diff
@@ -378,6 +414,16 @@ namespace parallel
...
       Settings settings;
 
+      /**
+       * Partitioner used in copy_triangulation().
+       */
+      std::function<void(dealii::Triangulation<dim, spacedim> &,
+                         const unsigned int)>
+        partitioner = [](dealii::Triangulation<dim, spacedim> &tria,
+                         const unsigned int                    n_partitions) {
+          GridTools::partition_triangulation_zorder(n_partitions, tria);
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: Initializers at class scope have some funny semantics, and we've generally avoided them. Just put a call to `set_partitioner()` into the constructor of the class with this lambda function.

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 20]
```diff
@@ -322,6 +325,39 @@ namespace parallel
...
+       *                    a reference to the triangullation to be partitioned
+       *                    and the number of partitions to be created.
+       * @param settings See the description of the Settings enumerator.
+       *
+       * @note As a default, GridTools::partition_triangulation_zorder() is used
+       *       as partitioner and data structures on multigrid levels are not
+       *       set up.
+       */
+      void
+      register_partitioner(
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: "register" has some connotation that the object lives outside the class, but here you're just copying it. Just call it 
```suggestion
      set_partitioner(
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 21]
```diff
@@ -322,6 +325,39 @@ namespace parallel
...
 
+      /**
+       * Implementation of the same function as in the base class.
+       *
+       * @param other_tria The triangulation to be copied. It can be a serial
+       *        Triangulation or a parallel::distributed::Triangulation. Both
+       *        can have been refined already.
+       *
+       * @note This function uses the partitioner registered with
+       *       register_partitioner().
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]: See below:
```suggestion
       *       set_partitioner().
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 22]
```diff
@@ -322,6 +325,39 @@ namespace parallel
...
+       *                    and the number of partitions to be created.
+       * @param settings See the description of the Settings enumerator.
+       *
+       * @note As a default, GridTools::partition_triangulation_zorder() is used
+       *       as partitioner and data structures on multigrid levels are not
+       *       set up.
+       */
+      void
+      register_partitioner(
+        std::function<void(dealii::Triangulation<dim, spacedim> &,
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
        const std::function<void(dealii::Triangulation<dim, spacedim> &,
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 23]
```diff
@@ -322,6 +325,39 @@ namespace parallel
...
+                           const unsigned int)> partitioner,
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
                           const unsigned int)> &partitioner,
```

File: include/deal.II/distributed/fully_distributed_tria.h [hunk 24]
```diff
@@ -322,6 +325,39 @@ namespace parallel
...
+        const Settings                          settings);
```
Comment (bangerth) [inline include/deal.II/distributed/fully_distributed_tria.h]:
```suggestion
        const Settings                          &settings);
```

File: include/deal.II/distributed/fully_distributed_tria.h (hunk 19 above)
Comment (bangerth) [reply include/deal.II/distributed/fully_distributed_tria.h]: Or use a member initializer.

File: source/distributed/fully_distributed_tria.cc [hunk 25]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+      // partitioned
+      if (dynamic_cast<const dealii::parallel::TriangulationBase<dim, spacedim>
+                         *>(&other_tria) == nullptr)
+        {
+          serial_tria =
+            std::make_unique<dealii::Triangulation<dim, spacedim>>();
+
+          // actually copy the serial triangulation
+          serial_tria->copy_triangulation(
+            *const_cast<dealii::Triangulation<dim, spacedim> *>(&other_tria));
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]: Why do you need the `const_cast`? The `copy_triangulation` function takes a `const` reference.

File: source/distributed/fully_distributed_tria.cc [hunk 26]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+    {
+      // pointer to the triangulation for which the construction data
+      // should be created (normally it is the input triangulation but
+      // in the case of a serial triangulation we create a copy which should
+      // be used)
+      const dealii::Triangulation<dim, spacedim> *other_tria_ptr = &other_tria;
+
+      // temporary serial triangulation (since the input triangulation is const
+      // and we might modify its subdomain_ids
+      // and level_subdomain_ids during partitioning); it is filled only
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
      // and level_subdomain_ids during partitioning); this pointer only points to anything
```

File: source/distributed/fully_distributed_tria.cc [hunk 27]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+      // if the source triangulation is serial
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
      // if the source triangulation is serial, and ensures that our copy is eventually deleted
```

File: source/distributed/fully_distributed_tria.cc [hunk 28]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+          // partition multigrid levels
+          if (this->is_multilevel_hierarchy_constructed())
+            GridTools::partition_multigrid_levels(*serial_tria);
+
+          // use the new serial triangulation to create the construction data
+          other_tria_ptr = serial_tria.get();
+        }
+
+      // create construction data
+      const auto construcion_data =
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
      const auto construction_data =
```

File: source/distributed/fully_distributed_tria.cc [hunk 29]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+        Utilities::create_construction_data_from_triangulation(
+          *other_tria_ptr,
+          this->mpi_communicator,
+          this->is_multilevel_hierarchy_constructed());
+
+      // finally create triangulation
+      this->create_triangulation(construcion_data);
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
      this->create_triangulation(construction_data);
```

File: source/distributed/fully_distributed_tria.cc [hunk 30]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::register_partitioner(
+      std::function<void(dealii::Triangulation<dim, spacedim> &,
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
      const std::function<void(dealii::Triangulation<dim, spacedim> &,
```

File: source/distributed/fully_distributed_tria.cc [hunk 31]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+                         const unsigned int)> partitioner,
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
                         const unsigned int)> &partitioner,
```

File: source/distributed/fully_distributed_tria.cc [hunk 32]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+      const Settings                          settings)
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
      const Settings                          &settings)
```

File: source/distributed/fully_distributed_tria.cc [hunk 33]
```diff
@@ -285,6 +289,74 @@ namespace parallel
...
+
+      // finally create triangulation
+      this->create_triangulation(construcion_data);
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::register_partitioner(
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc]:
```suggestion
    Triangulation<dim, spacedim>::set_partitioner(
```

**Comment by peterrum** [comment]
@bangerth No worries! I was pressuring in the other PR because it contains significant changes preventing (or making it at least difficult to add) additions of any modifications to `p:d:t`.

I have incorporated your suggestions! Thanks!

**Comment by bangerth** [comment]
OK to merge.

You might want to think about the following for a follow-up: It seems strange that the `set_partitioner()` function takes not only a partitioner, but also a `Setting` object. In all other triangulation classes, these settings are set once and for all in the constructor, and I suspect that you could do the same here. That `set_partitioner` gets the settings is particularly funny because the settings aren't actually given to the partitioner itself. So there is a certain level of cognitive dissonance here, and you might want to think about resolving this one way or the other.

**Comment by peterrum** [comment]
@bangerth I guess you are right. I'll do that in follow up PR.
